x/pairing/keeper/unresponsive_provider.go

```diff
@@ -0,0 +1,314 @@
+		}
+		currentEpoch = previousEpoch
+	}
+
+	// Go over previous epochs to count the CU serviced by the provider
+	epochAfterCountingProviderCu, err := k.countCuForUnresponsiveProviderAndComplainers(ctx, true, currentEpoch, epochsNumToCheckCUForUnresponsiveProvider, providersPunishEntryMap)
+	if err != nil {
+		return nil, utils.LavaError(ctx, k.Logger(ctx), "count_cu_unresponsive_provider", map[string]string{"err": err.Error()}, "couldn't count the CU of the unresponsive provider")
+	}
+
+	// Go over previous epochs (from the point the last loop finished) to count the CU of clients that complained about the provider
+	_, err = k.countCuForUnresponsiveProviderAndComplainers(ctx, false, epochAfterCountingProviderCu, epochsNumToCheckCUForComplainers, providersPunishEntryMap)
+	if err != nil {
+		return nil, utils.LavaError(ctx, k.Logger(ctx), "count_cu_complainers", map[string]string{"err": err.Error()}, "couldn't count the CU of the complainers")
+	}
+
+	// Get providers that should be punished
+	for punishEntryKey, punishEntry := range providersPunishEntryMap {
+		// provider doesn't have payments (serviced CU = 0)
+		if punishEntry.providerServicedCu == 0 {
+			chainID := strings.Split(punishEntryKey, "_")[1]
+
+			// Try getting the provider's stake entry
+			_, foundStakeEntry := k.epochStorageKeeper.GetEpochStakeEntries(ctx, currentEpoch, epochstoragetypes.ProviderKey, chainID)
+
+			// if stake entry is not found -> this is the provider's first epoch (didn't get a chance to stake yet) -> don't punish him
+			if !foundStakeEntry {
+				continue
+			}
+		}
+
+		// The provider served less CU than the sum of the complainers CU -> should be punished
+		if punishEntry.complainersUsedCu > punishEntry.providerServicedCu {
+			providersToPunishMap[punishEntryKey] = punishEntry.providerPaymentStorageIndexList
+		}
+	}
+
+	return providersToPunishMap, nil
+}
+
+// Function to count the CU serviced by the unresponsive provider or the CU of the complainers.
+// The function counts CU from <epoch> back. The number of epochs to go back is <epochsNumToGoBack>. The function returns the last visited epoch
+func (k Keeper) countCuForUnresponsiveProviderAndComplainers(ctx sdk.Context, isProvider bool, epoch uint64, epochsNumToGoBack uint64, providersPunishEntryMap map[string]punishEntry) (uint64, error) {
+	epochTemp := epoch
+
+	for counter := uint64(0); counter < epochsNumToGoBack; counter++ {
+		// Get previous epoch (from epochTemp)
+		previousEpoch, err := k.epochStorageKeeper.GetPreviousEpochStartForBlock(ctx, epochTemp)
+		if err != nil {
+			return 0, utils.LavaError(ctx, k.Logger(ctx), "get_previous_epoch", map[string]string{"err": err.Error(), "epoch": fmt.Sprintf("%+v", epochTemp)}, "couldn't get previous epoch")
+		}
+
+		// Get the current stake storages (from all chains)
+		stakeStorageList := k.getCurrentProviderStakeStorageList(ctx)
```

Comment:
create a function outside of this one that fetches all the staked providers of the CURRENT EPOCH, before the recommendedEpochNumToCollectPayment. after that create 1 function that goes over and handles unresponsiveness for 1 provider and do it for all providers

Reply:
done